modules/repository-gcs/src/test/java/org/elasticsearch/repositories/gcs/GoogleCloudStorageRetryingInputStreamTests.java

```diff
@@ -0,0 +1,148 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.repositories.gcs;
+
+import com.google.api.client.http.HttpRequest;
+import com.google.api.client.http.HttpResponse;
+import com.google.api.client.http.HttpTransport;
+import com.google.api.client.http.LowLevelHttpRequest;
+import com.google.api.client.http.LowLevelHttpResponse;
+import com.google.api.client.testing.http.HttpTesting;
+import com.google.api.client.testing.http.MockHttpTransport;
+import com.google.api.client.testing.http.MockLowLevelHttpRequest;
+import com.google.api.client.testing.http.MockLowLevelHttpResponse;
+import com.google.api.gax.retrying.RetrySettings;
+import com.google.cloud.storage.BlobId;
+import com.google.cloud.storage.StorageException;
+import com.google.cloud.storage.StorageOptions;
+
+import org.elasticsearch.common.io.Streams;
+import org.elasticsearch.core.Nullable;
+import org.elasticsearch.repositories.blobstore.RequestedRangeNotSatisfiedException;
+import org.elasticsearch.rest.RestStatus;
+import org.elasticsearch.test.ESTestCase;
+
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.Locale;
+
+import static org.hamcrest.Matchers.equalTo;
+import static org.hamcrest.Matchers.instanceOf;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+public class GoogleCloudStorageRetryingInputStreamTests extends ESTestCase {
+
+    private static final String BUCKET_NAME = "test-bucket";
+    private static final String BLOB_NAME = "test-blob";
+
+    public void testReadWithinBlobLength() throws IOException {
+        byte[] bytes = randomByteArrayOfLength(randomIntBetween(1, 512));
+        int position = randomIntBetween(0, Math.max(0, bytes.length - 1));
+        int maxLength = bytes.length - position;
+        int length = randomIntBetween(maxLength + 1, Integer.MAX_VALUE - 1);
+        try (var stream = createRetryingInputStream(bytes, randomIntBetween(1, 3), position, length)) {
```

Comment:
position and length are always non null here, should we randomize this?

Reply:
position and length are already randomized just few lines above.
(If I misread your comment and you meant something else, please rephrase)

Reply:
Sorry if I wasn't clear.

We're only testing reads with a position and a length, if we were passing null we could also test the `GoogleCloudStorageRetryingInputStream(Storage client, BlobId blobId)` default constructor I think.

Also, we're not checking the bytes that are effectively read, I think we should.

Reply:
Default constructor is not handy for unit testing because of its static init inside. I added one more (just for testing) which allows to plug mock object and keep semantics of defaul constructor `start: 0, end: Long.MAX_VALUE - 1`

Added a check for return bytes against requested